llvm/include/llvm/Support/GenericDomTreeConstruction.h

```diff
@@ -60,6 +60,10 @@ struct SemiNCAInfo {
   static constexpr bool IsPostDom = DomTreeT::IsPostDominator;
   using GraphDiffT = GraphDiff<NodePtr, IsPostDom>;
 
+  template <typename T>
+  using has_number_t =
+      decltype(GraphTraits<T *>::getNumber(std::declval<T *>()));
```

Comment:
You can expose the boolean value outside of this class to avoid using `is_detected` every time you use this.
```c++
template <typename T>
constexpr bool HasBlockNumber = is_detected<has_number_t<...>>;
```

Reply:
Also, you can add 'unit tests' for this detector (using `static_assert`).

Reply:
Good idea. It would make sense to have this in GraphTraits.h, with static_asserts in the specializations. I used `GraphHasNodeNumbers` for now, but I don't really like the naming. `GraphTraits<NodeT *>::HasNodeNumbers would (imho) be better, but would require changing every specialization... What do you think?

Reply:
That would also make sense to me.